File: ObjC/CGPathUnionObjC/CGPathUnionObjC/ShapesView.m
//
//  ShapesView.m
//  CGPathUnionObjC
//
//  Created by Don Mag on 6/30/23.
//

#import "ShapesView.h"

@interface ShapesView ()
{
	NSMutableArray <MyShape *>*theShapes;
	CAShapeLayer *shapeLayer;
	NSInteger activeIDX;
	CGPoint currentPT;
}
@end

@implementation ShapesView

- (instancetype)init
{
	self = [super init];
	if (self) {
		[self commonInit];
	}
	return self;
}
- (instancetype)initWithFrame:(CGRect)frame
{
	self = [super initWithFrame:frame];
	if (self) {
		[self commonInit];
	}
	return self;
}
- (instancetype)initWithCoder:(NSCoder *)coder
{
	self = [super initWithCoder:coder];
	if (self) {
		[self commonInit];
	}
	return self;
}

- (void)commonInit {
	shapeLayer = [CAShapeLayer new];
	shapeLayer.fillColor = UIColor.clearColor.CGColor;
	shapeLayer.strokeColor = UIColor.blueColor.CGColor;
	shapeLayer.lineWidth = 8.0;
	shapeLayer.lineJoin = kCALineJoinRound;
	[self.layer addSublayer:shapeLayer];
	
	theShapes = [NSMutableArray new];
	activeIDX = -1;
	currentPT = CGPointZero;
}

- (void)addShape: (MyShape *)shape {
	[theShapes addObject:shape];
	[self setNeedsLayout];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
	[super touchesBegan:touches withEvent:event];
	UITouch *touch = [touches anyObject];
	currentPT = [touch locationInView:self];
	activeIDX = -1;
	for (int i = 0; i < theShapes.count; i++) {
		MyShape *shape = theShapes[i];
		CGAffineTransform t = CGAffineTransformMakeTranslation(-shape.offset.x, -shape.offset.y);
		if (CGPathContainsPoint(shape.thePath, &t, currentPT, NO)) {
			activeIDX = i;
			break;
		}
	}
}
- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
	[super touchesMoved:touches withEvent:event];
	if (activeIDX == -1) { return; }
	UITouch *touch = [touches anyObject];
	CGPoint pt = [touch locationInView:self];
	CGPoint newOffset = theShapes[activeIDX].offset;
	newOffset.x += (pt.x - currentPT.x);
	newOffset.y += (pt.y - currentPT.y);
	theShapes[activeIDX].offset = newOffset;
	currentPT = pt;
	[self setNeedsLayout];
}

- (void)layoutSubviews {
	[super layoutSubviews];
	
	shapeLayer.path = nil;
	
	if (theShapes.count == 0) {
		return;
	}
	
	CGPathRef pth = nil;
	for (MyShape *shape in theShapes) {
		// move the path based on the offset property of MyPath
		CGAffineTransform t = CGAffineTransformMakeTranslation(shape.offset.x, shape.offset.y);
		CGPathRef pr = CGPathCreateCopyByTransformingPath(shape.thePath, &t);
		if (pth == nil) {
			// this is the first path
			pth = pr;
		} else {
			// union the paths
			pth = CGPathCreateCopyByUnioningPath(pth, pr, NO);
		}
	}
	shapeLayer.path = pth;
}

@end

File: Swift/CGPathUnionSwift/CGPathUnionSwift/ShapesView.swift
//
//  ShapesView.swift
//  CGPathUnionSwift
//
//  Created by Don Mag on 6/30/23.
//

import UIKit

class ShapesView: UIView {

	private var theShapes: [MyShape] = []
	private let shapeLayer: CAShapeLayer = CAShapeLayer()
	private var activeIDX: Int = -1
	private var currentPT: CGPoint = .zero
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		commonInit()
	}
	required init?(coder: NSCoder) {
		super.init(coder: coder)
		commonInit()
	}
	func commonInit() {
		shapeLayer.fillColor = UIColor.clear.cgColor
		shapeLayer.strokeColor = UIColor.red.cgColor
		shapeLayer.lineWidth = 8.0
		shapeLayer.lineJoin = .round
		layer.addSublayer(shapeLayer)
	}
	
	func addShape(_ shape: MyShape) {
		theShapes.append(shape)
		setNeedsLayout()
	}

	override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
		super.touchesBegan(touches, with: event)
		guard let t = touches.first else { return }
		currentPT = t.location(in: self)
		activeIDX = -1
		for i in 0..<theShapes.count {
			let shape = theShapes[i]
			if shape.thePath.contains(currentPT, transform: .init(translationX: -shape.offset.x, y: -shape.offset.y)) {
				activeIDX = i
				break
			}
		}
	}
	override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
		super.touchesMoved(touches, with: event)
		guard activeIDX != -1, let t = touches.first else { return }
		let pt = t.location(in: self)
		theShapes[activeIDX].offset.x += (pt.x - currentPT.x)
		theShapes[activeIDX].offset.y += (pt.y - currentPT.y)
		currentPT = pt
		setNeedsLayout()
	}
	
	override func layoutSubviews() {
		super.layoutSubviews()
		var pth: CGPath!
		theShapes.forEach { shape in
			var t = CGAffineTransform(translationX: shape.offset.x, y: shape.offset.y)
			if let pr = shape.thePath.mutableCopy(using: &t) {
				if pth == nil {
					pth = pr
				} else {
					pth = pth.union(pr)
				}
			}
			//pth.addPath(shape.thePath, transform: .init(translationX: shape.offset.x, y: shape.offset.y))
		}
		shapeLayer.path = pth
//		if #available(iOS 116.0, *) {
//			shapeLayer.path = pth.normalized()
//		} else {
//			shapeLayer.path = pth
//		}
	}
	
}
